### PR TITLE
Integration of all parts

### LINKED ISSUE #80: Setup monorepo to integrate the navigator
https://github.com/mamba-org/jupyter_conda/pull/79 brings the navigator from https://github.com/mamba-org/mamba-navigator to the repo.

The next step would be to use a monorepo structure to be able to reuse the frontend components easily between the lab and the standalone app.

### LINKED ISSUE #85: Package the navigator
## Description

Originally tracked in https://github.com/mamba-org/mamba-navigator/issues/13

So the navigator can be installed via `pip`.

### PR DESCRIPTION
**Pull request by fcollonval**
This is a major refractoring of the code for integrating the extensions and the navigator as well as publishing them.

- [x] Setup multipackages repo
- [x] Fixes #80 
- [x] Fixes #85 
- [x] Migrate tests
- [x] Package the nbextension
- [x] Update setup scripts
- [x] Update README
- [x] Update binder
- [ ] Update publication workflow

### CODE REVIEW

**Comment by fcollonval** [comment]
Binder link:
https://mybinder.org/v2/gh/mamba-org/jupyter_conda/integration?urlpath=lab

**Comment by fcollonval** [comment]
@jtpio This is ready for review.

The new structure is as follow:

- All backend code in `jupyter_conda`
- Frontend code split in 4 parts:
  - packages/common : Common code for the lab extension and the navigator
  - packages/labextension: Lab extension  
    Published as npm package in jupyter_conda/labextension
  - packages/navigator: Navigator frontend  
    Pubished as static files in jupyter_conda/navigator/static
  - jupyter_conda/nbextension: Classical notebook extension

The navigator can be launch from an entry point `navigator` or as the python package main `python -m jupyter_conda`

Questions:
- What naming for the different packages? For now
  - @mamba-org/common
  - @mamba-org/conda-lab
  - @mamba-org/navigator (no public publication)
  - jupyter_conda for Python code
- What logic for versioning and publication?

I think we can discuss that with Sylvain in a call.

**Comment by fcollonval** [comment]
I did a quick analysis of the navigator bundle:
image

We got the blueprintjs of course... But more surprising is `codemirror`. I realized that `@jupyterlab/application` implies `codemirror` due to `@jupyterlab/rendermime`. This is used only to render code in markdown (https://github.com/jupyterlab/jupyterlab/blob/2.2.x/packages/rendermime/src/renderers.ts#L809). This makes me wondering if there is a space for modularity improvement.

Back to this code, switching from `react-virtualized` to `react-window` should really be considered #77.

**Comment by jtpio** [comment]
Wow thanks a lot for tacking this!

> What logic for versioning and publication?

Lab does it by keeping versions in sync in all packages, even if there were no real change. Maybe it could make things simpler that way. 

> This is used only to render code in markdown (https://github.com/jupyterlab/jupyterlab/blob/2.2.x/packages/rendermime/src/renderers.ts#L809). This makes me wondering if there is a space for modularity improvement in jlab core.

Looks like this could indeed be open as an issue on the core repo (if there isn't one already).

File: README.md [hunk 1]
```diff
@@ -38,11 +35,40 @@ jupyter labextension install jupyterlab_conda
 > Optionally, you could install [`jupyterlab-tour`](https://github.com/fcollonval/jupyterlab-tour) to
 > add a help tour for the conda packages manager.
 
+## JupyterLab
+
+This extension add a new entry _Conda Packages Manager_ in the _Settings_ menu.
```
Comment (jtpio) [inline README.md:40]:
```suggestion
This extension adds a new entry _Conda Packages Manager_ in the _Settings_ menu.
```

File: jupyter_conda/_version.py [hunk 2]
```diff
@@ -1,2 +1,2 @@
-version_info = (3, 4, 1)
+version_info = (4, 0, 0)
```
Comment (jtpio) [inline jupyter_conda/_version.py:1]: It's probably a matter of style and preference, but should the version bump be done in this PR or in a separate release commit?

File: README.md [hunk 3]
```diff
@@ -92,70 +109,34 @@ There are three ways to create an environment:
 ## Development
 
 ```shell
-conda create -y -n jupyter_conda python jupyterlab~=2.1
-conda install -y -n jupyter_conda --file requirements_dev.txt -c conda-forge
+conda create -c conda-forge -y -n jupyter_conda python jupyterlab=2
+conda install-c conda-forge -y -n jupyter_conda --file requirements_dev.txt
```
Comment (jtpio) [inline README.md:113]:
```suggestion
conda install -c conda-forge -y -n jupyter_conda --file requirements_dev.txt 
```

**Comment by jtpio** [comment]
Looks like the builds fails on Binder with:

image

File: packages/navigator/package.json [hunk 4]
```diff
@@ -70,7 +68,8 @@
     "typescript": "~3.9.0",
     "url-loader": "~3.0.0",
     "watch": "~1.0.2",
-    "webpack": "5.0.0-beta.21",
+    "webpack": "^5.0.0",
```
Comment (jtpio) [inline packages/navigator/package.json:71]: We can probably stick to `"webpack": "^4.0.0"` for now so it uses the same version as in lab 2

Comment (jtpio) [reply packages/navigator/package.json:71]: And then bump to `^5.0.0` when updating to lab 3.

**Comment by jtpio** [comment]
Just tried locally with this change https://github.com/mamba-org/jupyter_conda/pull/91#discussion_r507752583:

```diff
diff --git a/packages/navigator/package.json b/packages/navigator/package.json
index 95b519e..28fca2a 100644
--- a/packages/navigator/package.json
+++ b/packages/navigator/package.json
@@ -68,7 +68,7 @@
     "typescript": "~3.9.0",
     "url-loader": "~3.0.0",
     "watch": "~1.0.2",
-    "webpack": "^5.0.0",
+    "webpack": "^4.0.0",
     "webpack-bundle-analyzer": "^3.9.0",
     "webpack-cli": "^3.3.10",
     "whatwg-fetch": "^3.0.0"
```

And it seems to be working nicely. Thanks again!

File: jupyter_conda/_version.py (hunk 2 above)
Comment (fcollonval) [reply jupyter_conda/_version.py:1]: This is a good idea.

File: packages/navigator/package.json (hunk 4 above)
Comment (fcollonval) [reply packages/navigator/package.json:71]: Indeed I agree, it will be nice to release first in the current Jupyter environment. Then switch to JLab3.

**Comment by fcollonval** [comment]
> 
> 
> Looks like the builds fails on Binder with:
> 
> image

Strange I just succeeded to launch it (OVH instance).

**Pull request merged**